### PR TITLE
Update Salt to work with Jinja >= and <= 3.1.0

### LINKED ISSUE #61848: [BUG] Jinja 3.1 moved Markup
**Description**
Jinja 3.1 was [released today](https://jinja.palletsprojects.com/en/3.1.x/changes/) and removed some previously deprecated code. At minimum, this breaks [this line](https://github.com/saltstack/salt/blob/master/salt/utils/jinja.py#L29) in `salt/utils/jinja.py` as `Markup` should now be imported from `MarkupSafe`:

```py
from jinja2 import BaseLoader, Markup, TemplateNotFound, nodes
```

The [base Jinja requirement is currently not pinned](https://github.com/saltstack/salt/blob/master/requirements/base.txt#L1), so we'll either need to require <3.1 or change the import and require >=3.1 (or narrow the pin even further).

### PR DESCRIPTION
**Pull request by Ch3LL**
### What does this PR do?
Updates Markup and contextfunction to work with all versions of Jinja2. Markup was moved to markupsafe and contextfunction was changed to pass_context in Jinja 3.1

### What issues does this PR fix or reference?
Fixes: https://github.com/saltstack/salt/issues/61848 https://github.com/saltstack/salt/issues/60745

### CODE REVIEW

File: salt/utils/jinja.py [hunk 1]
```diff
@@ -35,6 +35,12 @@
 from salt.utils.odict import OrderedDict
 from salt.utils.versions import LooseVersion
 
+try:
+    from jinja2 import Markup
+except ImportError:
+    # Markup moved to markupsafe in jinja>= 3.1
+    from markupsafe import Markup
+
```
Comment (s0undt3ch) [inline salt/utils/jinja.py:43]:
```suggestion
try:
    from markupsafe import Markup
except ImportError:
    # jinja < 3.1
    from jinja2 import Markup

```

File: salt/utils/jinja.py [hunk 2]
```diff
@@ -707,7 +713,13 @@ def method_call(obj, f_name, *f_args, **f_kwargs):
     return getattr(obj, f_name, lambda *args, **kwargs: None)(*f_args, **f_kwargs)
 
 
-@jinja2.contextfunction
+if jinja2.__version__ < "3.0.0":
+    contextfunction = jinja2.contextfunction
+else:
+    contextfunction = jinja2.pass_context
```
Comment (s0undt3ch) [inline salt/utils/jinja.py:719]:
```suggestion
try:
    contextfunction = jinja2.pass_context
except AttributeError:
    contextfunction = jinja2.contextfunction
```

**Review by s0undt3ch — CHANGES_REQUESTED** [review]

**Review by twangboy — DISMISSED** [review]

**Review by s0undt3ch — DISMISSED** [review]

**Comment by krionbsd** [comment]
re-run all

**Review by waynew — DISMISSED** [review]

**Comment by myii** [comment]
I've used this patch when building the `master` branch images for this week's testing across the Formulas' org -- all seems to be working well, thanks.

**Comment by myii** [comment]
Perhaps this will also close #60745 -- I haven't tested that.

**Comment by MKLeb** [comment]
re-run all

**Review by MKLeb — APPROVED** [review]

**Review by garethgreenaway — APPROVED** [review]

**Review by waynew — APPROVED** [review]

**Comment by jpic** [comment]
Great! Is this going to be backported on older salt versions?

**Pull request merged**